File: DotNetGotchas/VBDN/PolymorphismTooSoon/PolymorphismAndConstruction/Manager.vb
'Manager.vb
Public Class Manager
    Inherits Employee
    Private theRoom As Room = Nothing
    Private managementLevel As Integer = 0

    Public Sub New(ByVal level As Integer)
        Console.WriteLine("Manager's constructor called")

        managementLevel = level

        If level < 2 Then
            theRoom = New Room
        Else
            theRoom = New ExecutiveRoom
        End If
    End Sub
    Public Overrides Sub Work()
        Console.WriteLine("Manager's work called")

        theRoom.OpenWindow()
        MyBase.Work()
    End Sub
End Class

File: DotNetGotchas/CSharp/PolymorphismTooSoon/PolymorphismAndConstruction/Manager.cs
//Manager.cs
using System;

namespace ProblemPolymorphismConstruction
{
	public class Manager : Employee
	{
		private Room theRoom = null;
		private int managementLevel = 0;

		public Manager(int level)
		{
			Console.WriteLine("Manager's constructor called");

			managementLevel = level;

			if (level < 2)
                theRoom = new Room();
			else
				theRoom = new ExecutiveRoom();
		}

		public override void Work()
		{
			Console.WriteLine("Manager's work called");

			theRoom.OpenWindow();
			base.Work();
		}
	}
}
